stablehlo/dialect/TypeInference.cpp

```diff
@@ -2670,10 +2670,27 @@ LogicalResult inferSelectOp(
                                "requires the same shape for all operands");
 
   // The output shape should be derived from the most specific parts of the
-  // `onTrue` and `onFalse` (see documentation for details).
-  SmallVector<Type> inferredReturnTypes;
-  return inferMostSpecificTypeComponents(location, {trueType, falseType},
-                                         inferredReturnShapes);
+  // `onTrue` and `onFalse` when the bounds are missing (see documantation for
+  // `inferMostSpecificTypeComponents` for details).
+  auto onTrueRankedTy = onTrue.getType().dyn_cast<RankedTensorType>();
+  auto onFalseRankedTy = onFalse.getType().dyn_cast<RankedTensorType>();
+  if (!onTrueRankedTy || !onFalseRankedTy ||
+      (encodingToBounds(onTrueRankedTy.getEncoding()).empty() &&
+       encodingToBounds(onFalseRankedTy.getEncoding()).empty()))
+    return inferMostSpecificTypeComponents(location, {trueType, falseType},
+                                           inferredReturnShapes);
+
+  // The output shape and bounds are derived using the least specifc rules for
+  // `onTrue` and `onFalse` when the bounds are present (see documentation for
+  // `inferLeastSpecificType` for details).
+  auto inferredTypeOrErr =
+      inferLeastSpecificType(location, {onTrueRankedTy, onFalseRankedTy});
```

Comment:
Also, why are we using inferLeastSpecificType for SelectOp?

Perhaps, that's because CaseOp/IfOp use inferLeastSpecificType, and both perform operations conditionally? If that's what you had in mind, then I think that SelectOp is materially different.

For CaseOp/IfOp, execution can take one of the several available paths, so it makes sense to use inferLeastSpecificType, because we don't know in advance which path will be taken. Also, in many other programming languages, conditional expressions use least upper bounds for type inference. That checks out.

For SelectOp, execution can take only one available path. Exactly the same as other elementwise ops, the op requires `on_true` and `on_false` to have the same shape at runtime because it needs to zip them together. Given that, I don't think there's a case here to deviate from inferMostSpecificType that's used in other elementwise ops.

Reply:
Hmm. This claim changed my mind (take back my previous approval, sorry for this confusion).
`inferLeastSpecificType` is not proper for `select`, For example: 
If lhs is `tensor<4xi32>` and rhs is `tensor<?xi32>`, then the result have to be `tensor<4xi32>` instead of any other form, so as rhs and `pred`:  If any dim is larger or smaller than 4, then the semantic meaning of this op does not make sense anymore.
If both lhs and rhs are `tensor<?xi32>` but have difference bounds, then the result must be align with the operand with smaller the bound.
So we need stick with `inferMostSpecificType` to infer shape (not element type) from all `pred` (if rank(pred) 
 !=0), `lhs`, `rhs`.

Reply:
Makes sense, thanks for the input.